src/k8s/pkg/client/kubernetes/upgrades.go

```diff
@@ -0,0 +1,178 @@
+package kubernetes
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/canonical/k8s/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+// TODO: If the upgrade CRD grows, consider using kubebuilder.
+const (
+	UpgradePhaseNodeUpgrade    = "NodeUpgrade"
+	UpgradePhaseFeatureUpgrade = "FeatureUpgrade"
+	UpgradePhaseFailed         = "Failed"
+	UpgradePhaseCompleted      = "Completed"
+)
+
+const (
+	kind    = "Upgrade"
+	group   = "k8sd.io"
+	version = "v1alpha"
+)
+
+type UpgradeMetadata struct {
+	Name string `json:"name,omitempty"`
+}
+
+type UpgradeStatus struct {
+	Phase         string   `json:"phase,omitempty"`
+	UpgradedNodes []string `json:"upgradedNodes,omitempty"`
+}
+
+type Upgrade struct {
+	APIVersion string          `json:"apiVersion,omitempty"`
+	Kind       string          `json:"kind,omitempty"`
+	Metadata   UpgradeMetadata `json:"metadata,omitempty"`
+	Status     UpgradeStatus   `json:"status,omitempty"`
+}
+
+func NewUpgrade(name string) Upgrade {
+	return Upgrade{
+		APIVersion: group + "/" + version,
+		Kind:       kind,
+		Metadata:   UpgradeMetadata{Name: name},
+		Status:     UpgradeStatus{Phase: UpgradePhaseNodeUpgrade, UpgradedNodes: []string{}},
+	}
+}
+
+func (c *Client) k8sdIoRestClient() (*rest.RESTClient, error) {
+	k8sdConfig := c.RESTConfig()
+	k8sdConfig.GroupVersion = &schema.GroupVersion{Group: group, Version: version}
+	k8sdConfig.NegotiatedSerializer = serializer.NewCodecFactory(runtime.NewScheme())
+
+	restClient, err := rest.RESTClientFor(k8sdConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+	return restClient, nil
+}
+
+// GetInProgressUpgrade returns the upgrade CR that is currently in progress.
+// TODO(ben): Maybe make this more generic, e.g. GetUpgrade(filterFunc func(Upgrade) bool) (*Upgrade, error)
+func (c *Client) GetInProgressUpgrade(ctx context.Context) (*Upgrade, error) {
+	log := log.FromContext(ctx).WithValues("upgrades", "GetInProgressUpgrade")
+
+	restClient, err := c.k8sdIoRestClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+
+	upgrades, err := restClient.Get().AbsPath(fmt.Sprintf("/apis/%s/%s/upgrades", group, version)).DoRaw(ctx)
```

Comment:
nit: Should we also turn `fmt.Sprintf("/apis/%s/%s/upgrades", group, version)` into a `const`?